### PR TITLE
added github action for continuous integration

### PR DESCRIPTION
**Pull request by fbergmann**
## Description
* added github workflow based on @alessandrofelder work for libSBML. 
* needed a number of changes to ensure that:
   *  namespaces would compile
   * expat
   * xerces 
   * would be working
   * builds on windows / ubuntu supported

## Motivation and Context
continuous integration will help quickly sorting out what works and what does not

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated all documentation necessary.
- [ ] I have checked spelling in (new) comments.

## Testing
- [X] Testing is done automatically 
- [ ] This cannot be tested automatically

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        with:
+          path: ${{runner.workspace}}/.ccache
+          key:
+            ${{ runner.os }}-test-${{ matrix.xml_parser_option }}-${{
+            matrix.with_namespace }}-${{ matrix.strict }}-${{
+            steps.ccache_cache_timestamp.outputs.timestamp }}
+          restore-keys: |
+            ${{ runner.os }}-test-${{ matrix.xml_parser_option }}-${{ matrix.with_namespace }}-${{ matrix.strict }}-${{ steps.ccache_cache_timestamp.outputs.timestamp }}
+            ${{ runner.os }}-test-${{ matrix.xml_parser_option }}-${{ matrix.with_namespace }}-${{ matrix.strict }}-
+            ${{ runner.os }}-test-${{ matrix.xml_parser_option }}-${{ matrix.with_namespace }}-${{ matrix.strict }}-
```
Comment (mattagape) [inline .github/workflows/build.yml:108]: Forgive my ignorance, but what are these identical lines doing?

File: src/liblx/xml/common/common.h [hunk 2]
```diff
@@ -70,9 +70,9 @@
  * into the XML produced by libSBML.  This is used by
  * StringBuffer_appendReal.
  */
-#define LIBSBML_FLOAT_FORMAT "%.15g"
+#define LIBLX_FLOAT_FORMAT "%.15g"
 
-#define LIBSBML_DOUBLE_PRECISION 15
+#define LIBLX_DOUBLE_PRECISION 15
 
 static const int SBML_INT_MAX = 2147483647;
```
Comment (mattagape) [inline src/liblx/xml/common/common.h:77]: Do we still want these listed as `SBML_` ?

File: src/liblx/xml/sbmlMemoryStubs.h [hunk 3]
```diff
@@ -9,12 +9,19 @@
...
+#include <cstddef>
 
 extern "C"
 char* safe_strdup(const char* s);
 extern "C"
 void* safe_malloc (size_t size);
 
+/**
+ * Easier-to-read and NULL-friendly string comparison.
+ */
```
Comment (mattagape) [inline src/liblx/xml/sbmlMemoryStubs.h:29]: This is duplicating some of my recent PR (although admittedly still under review)
Also, does it need to be declared as `extern "C"`?

**Review by mattagape — CHANGES_REQUESTED** [review]
Please look at the few comments I made. Thanks

File: .github/workflows/build.yml (hunk 1 above)
Comment (alessandrofelder) [reply .github/workflows/build.yml:108]: This is a relic from when I tried to use `ccache` that got copied over. It's a non-critical bug that has been on my todo for a while to remove/improve on. Can be removed here, I think.

File: src/liblx/xml/sbmlMemoryStubs.h (hunk 3 above)
Comment (fbergmann) [reply src/liblx/xml/sbmlMemoryStubs.h:29]: this whole file needs changing, see the exchange today. it should use the begin_c_decls and namespaces for everything. But i can add it ther.

File: src/liblx/xml/common/common.h (hunk 2 above)
Comment (fbergmann) [reply src/liblx/xml/common/common.h:77]: i could go with either definition, LIBLX is more consistent

File: CMakeLists.txt [hunk 4]
```diff
@@ -43,11 +43,10 @@ if(POLICY CMP0058)
...
 
 
-
-set(LIBLX_ROOT_SOURCE_DIR "${CMAKE_SOURCE_DIR}" CACHE PATH
-    "Path to the libSBML root source directory")
-set(LIBLX_ROOT_BINARY_DIR "${CMAKE_BINARY_DIR}" CACHE PATH
-    "Path to the libSBML root build directory")
+set(LIBLX_ROOT_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}" CACHE PATH
+    "Path to the libLX root source directory")
+set(LIBSLX_ROOT_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}" CACHE PATH
```
Comment (mattagape) [inline CMakeLists.txt]: Should this be `LIBLX_ROOT_BINARY_DIR` instead of `LIBSLX_ROOT_BINARY_DIR`?

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -1,10 +1,10 @@
 ###############################################################################
-# Description       : CMake build script for libSBML
+# Description       : CMake build script for libLX
 # Original author(s): Frank Bergmann <fbergman@caltech.edu>
 # Organization      : California Institute of Technology
 #
-# This file is part of libSBML.  Please visit http://sbml.org for more
-# information about SBML, and the latest version of libSBML.
+# This file is part of libLX.  Please visit http://sbml.org for more
+# information about LX, and the latest version of libLX.
```
Comment (mattagape) [inline src/CMakeLists.txt]: I think `LX` should be `libLX` here

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -37,24 +37,13 @@ endif()
 
 ###############################################################################
 #
-# Initialize SBML sources
+# Initialize LX sources
```
Comment (mattagape) [inline src/CMakeLists.txt]: `libLX`?

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -102,22 +91,76 @@ configure_file(
...
-file(GLOB header_files "${CMAKE_CURRENT_SOURCE_DIR}/liblx/xml/*.h")
-install(FILES ${header_files} DESTINATION include/liblx/xml)
+foreach(dir common compress xml)
+	file(GLOB header_files "${CMAKE_CURRENT_SOURCE_DIR}/liblx/xml/${dir}/*.h")
+	install(FILES ${header_files} DESTINATION include/liblx/xml/${dir})
+
+endforeach(dir)
+
+
+file(GLOB files "${CMAKE_CURRENT_SOURCE_DIR}/sbml/*.h")
```
Comment (mattagape) [inline src/CMakeLists.txt]: In these two lines (101 and 102) we need `liblx` not `sbml`, presumably

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -285,9 +328,9 @@ if(WITH_XERCES)
 
 endif(WITH_XERCES)
 
-# Add SBMLNamespaces until the reserved namespace mechanism is replaced
+# Add LXNamespaces until the reserved namespace mechanism is replaced
 #set(XML_SOURCES ${XML_SOURCES}
-#	liblx/xml/SBMLNamespaces.cpp
+#	liblx/xml/LXNamespaces.cpp
```
Comment (mattagape) [inline src/CMakeLists.txt]: Should this be `libLXNamespaces.cpp`?

File: src/bindings/python/CMakeLists.txt [hunk 9]
```diff
@@ -175,75 +82,19 @@ add_custom_command(
...
-        -g ${CMAKE_CURRENT_SOURCE_DIR}/../../../docs/src/common-graphics
-
-    COMMAND "${CMAKE_COMMAND}"
-    ARGS  -E rename pydoc-doxygen.i pydoc.i
-
     COMMAND "${SWIG_EXECUTABLE}"
     ARGS  -I${CMAKE_CURRENT_SOURCE_DIR}/../swig/
         -I${CMAKE_CURRENT_SOURCE_DIR} 
         -I${CMAKE_CURRENT_BINARY_DIR} 
         -I${LIBSBML_ROOT_BINARY_DIR}/src
```
Comment (mattagape) [inline src/bindings/python/CMakeLists.txt]: Lines in this block need to be `LIBLX_`, not `LIBSBML_`

File: src/bindings/python/CMakeLists.txt [hunk 10]
```diff
@@ -175,75 +82,19 @@ add_custom_command(
...
         -I${LIBSBML_ROOT_SOURCE_DIR}/src
         -I${LIBSBML_ROOT_SOURCE_DIR}/include
-        -DGENERATING_DOCS
         -c++
         -python    
         ${SWIG_EXTRA_ARGS}     
         -o ${CMAKE_CURRENT_BINARY_DIR}/libsbml_wrap.cpp
```
Comment (mattagape) [inline src/bindings/python/CMakeLists.txt]: This and next line need to be `liblx` not `libsbml`

File: src/bindings/python/CMakeLists.txt [hunk 11]
```diff
@@ -175,75 +82,19 @@ add_custom_command(
...
-        ${SWIG_EXTRA_ARGS}     
-        -o ${CMAKE_CURRENT_BINARY_DIR}/libsbml_wrap.cpp 
-        ${CMAKE_CURRENT_SOURCE_DIR}/libsbml.i
-
-    COMMAND "${CMAKE_COMMAND}"
-    ARGS  -E rename pydoc.i pydoc-normal.i 
-
     COMMAND "${CMAKE_COMMAND}"
     ARGS -DCUR_BIN_DIRECTORY=\"${CMAKE_CURRENT_BINARY_DIR}\"
          -DVERSION=\"${LIBSBML_DOTTED_VERSION}\"
```
Comment (mattagape) [inline src/bindings/python/CMakeLists.txt]: `LIBSBML_DOTTED_VERSION` -> `LIBLX_DOTTED_VERSION` ?

File: src/bindings/python/CMakeLists.txt [hunk 12]
```diff
@@ -256,9 +107,8 @@ add_custom_command(
          -P "${CMAKE_CURRENT_SOURCE_DIR}/patch-python.cmake"
 
     MAIN_DEPENDENCY ${CMAKE_CURRENT_SOURCE_DIR}/libsbml.i
```
Comment (mattagape) [inline src/bindings/python/CMakeLists.txt]: `liblx.i`

File: src/bindings/python/CMakeLists.txt [hunk 13]
```diff
@@ -256,9 +107,8 @@ add_custom_command(
...
+
     DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/CMakeLists.txt
-        ${CMAKE_CURRENT_SOURCE_DIR}/doc-converter/rewrite_pydoc.py
-        ${CMAKE_CURRENT_SOURCE_DIR}/../swig/swigdoc.py
         ${SWIG_DEPENDENCIES} ${LIBSBML_HEADER_FILES}
```
Comment (mattagape) [inline src/bindings/python/CMakeLists.txt]: `LIBLX_HEADER_FILES`

**Review by mattagape — CHANGES_REQUESTED** [review]
A few `LIBSBML` occurrences were not changed into `LIBLX` (etc.)

File: CMakeLists.txt (hunk 4 above)
Comment (fbergmann) [reply CMakeLists.txt]: yes

File: src/CMakeLists.txt (hunk 5 above)
Comment (fbergmann) [reply src/CMakeLists.txt]: the comments will need general revisions. sbml.org does not contain any information about liblx, and the documentation will reflect this later. I suggest leaving it till then.

Comment (mattagape) [reply src/CMakeLists.txt]: OK

**Comment by fbergmann** [comment]
closing this, as we only wanted the CI tests in this one, and it is conflicting with other branches. see #16

**Pull request closed without merging**